# HG changeset patch
# User Wohlstand <admin@wohlnet.ru>
# Date 1530492837 -10800
#      Mon Jul 02 03:53:57 2018 +0300
# Node ID 523e803a0192984b78e5d415fd11a84b461734f0
# Parent  bf70bfa0221537b342e4833a09ea223731402874
Fixed bug 4210 - SSE2-based converter makes junk result of S32 -> Float

At the HG state e604fe493d45, 64-bit assemblies are using SSE2-based resampler, produces junk sound when converting the S32 -> Float32 -> S16 chain. The `NEED_SCALAR_CONVERTER_FALLBACKS` thing works perfectly.

If I will find a reason that caused this mistake, I'll send a patch by myself.

diff -r bf70bfa02215 -r 523e803a0192 src/audio/SDL_audiotypecvt.c
--- a/src/audio/SDL_audiotypecvt.c	Sun Jul 01 19:50:00 2018 +0300
+++ b/src/audio/SDL_audiotypecvt.c	Mon Jul 02 03:53:57 2018 +0300
@@ -533,7 +533,7 @@
         const __m128i *mmsrc = (const __m128i *) src;
         while (i >= 4) {   /* 4 * sint32 */
             /* shift out lowest bits so int fits in a float32. Small precision loss, but much faster. */
-            _mm_store_ps(dst, _mm_mul_ps(_mm_cvtepi32_ps(_mm_srli_epi32(_mm_load_si128(mmsrc), 8)), divby8388607));
+            _mm_store_ps(dst, _mm_mul_ps(_mm_cvtepi32_ps(_mm_srai_epi32(_mm_load_si128(mmsrc), 8)), divby8388607));
             i -= 4; mmsrc++; dst += 4;
         }
         src = (const Sint32 *) mmsrc;
